source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -49,92 +58,125 @@ Http::FilterHeadersStatus CacheFilter::decodeHeaders(Http::RequestHeaderMap& hea
   lookup_ = cache_.makeLookupContext(std::move(lookup_request));
 
   ASSERT(lookup_);
-
   ENVOY_STREAM_LOG(debug, "CacheFilter::decodeHeaders starting lookup", *decoder_callbacks_);
-  lookup_->getHeaders([this](LookupResult&& result) { onHeaders(std::move(result)); });
-  if (state_ == GetHeadersState::GetHeadersResultUnusable) {
+
+  // Used exclusively to access the RequestHeaderMap in onHeaders
+  request_headers_ = &headers;
+  getHeaders();
+
+  if (get_headers_state_ == GetHeadersState::GetHeadersResultUnusable) {
     // onHeaders has already been called, and no usable cache entry was found--continue iteration.
     return Http::FilterHeadersStatus::Continue;
   }
   // onHeaders hasn't been called yet--stop iteration to wait for it, and tell it that we stopped
```

Comment:
I did not change the state management here at all, I only renamed the member from `state_` to `get_headers_state_`. The conditions and comments are untouched.
If onHeaders is called synchronously with a given response then returning `StopIteration` is appropriate because a cache hit was found so the request should not continue.
Do you think I should elaborate the comments?

Reply:
Thank you for reworking the state management--it seems much more clear now.